### PR TITLE
More efficient computation of recursive trees

### PR DESCRIPTION
**Pull request by ppedrot**
We try to be a bit less stupid in the expansion of rtrees both in the kernel (for guard checking) and outside (for the building of recursors). Instead of expanding nodes on sight, we try to delay as much as possible the moment we actually build the tree, as it may otherwise result in an exponential blowup for data that is not even needed. The new implementation is based on the efficient Esubst API, as it's basically a variant of term substitution for a simpler grammar.

Seriously mitigates #19650. The example is still taking ~2s on my machine and spending 99% of its time in rtree expansion, but this should be compared with the previous situation, which was taking ~80s. We can thus do better but this requires more in-depth fiddling with the guard checker, whereas the current PR is straightforward.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

File: kernel/rtree.mli [hunk 1]
```diff
@@ -54,6 +54,9 @@ val is_node : 'a t -> bool
 (** Destructors (recursive calls are expanded) *)
 val dest_node  : 'a t -> 'a * 'a t array array
 
+(** Fist projection of {!dest_node} *)
```
Comment (SkySkimmer) [inline kernel/rtree.mli:57]: fist?

Comment (ppedrot) [reply kernel/rtree.mli]: Fixed.

**Comment by SkySkimmer** [comment]
Bench said
~~~
┌─────────────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬─────────────────────────┐
│                                     │      user time [s]      │           CPU instructions            │  max resident mem [KB]  │
│                                     │                         │                                       │                         │
│            package_name             │   NEW      OLD    PDIFF │      NEW             OLD        PDIFF │   NEW      OLD    PDIFF │
├─────────────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼─────────────────────────┤
│               coq-mathcomp-solvable │   95.74    97.15  -1.45 │   670059675750    670080213780  -0.00 │  876184   878808  -0.30 │
│                             coq-vst │  856.10   865.13  -1.04 │  6494619974032   6495657555702  -0.02 │ 1795752  1796348  -0.03 │
│                 coq-metacoq-erasure │  517.63   522.86  -1.00 │  3590377477943   3635710696241  -1.25 │ 1710152  1720736  -0.62 │
│                   coq-iris-examples │  398.08   402.07  -0.99 │  2663596368437   2663828239797  -0.01 │ 1075288  1075096   0.02 │
│                        coq-bedrock2 │  327.72   330.85  -0.95 │  2774640827420   2774655628387  -0.00 │  862728   864684  -0.23 │
│                  coq-mathcomp-field │   92.47    93.21  -0.79 │   651180971608    651189868004  -0.00 │ 1202416  1201304   0.09 │
│                         coq-unimath │ 1472.61  1484.27  -0.79 │ 12182455489170  12182833742213  -0.00 │ 1126496  1121044   0.49 │
│                         coq-coqutil │   42.70    42.95  -0.58 │   270322167290    270387629945  -0.02 │  554716   557116  -0.43 │
│              coq-mathcomp-character │   78.30    78.73  -0.55 │   528235949004    528150275142   0.02 │  935340   935160   0.02 │
│                   coq-metacoq-pcuic │  676.32   679.41  -0.45 │  4413034513303   4457902417169  -1.01 │ 2343700  2075456  12.92 │
│          coq-performance-tests-lite │  901.34   904.81  -0.38 │  7309580013914   7309130005470   0.01 │ 2439272  2439200   0.00 │
│              coq-mathcomp-odd-order │  558.70   560.79  -0.37 │  3956953945886   3956590985150   0.01 │ 1572028  1578320  -0.40 │
│                        coq-coqprime │   51.95    52.11  -0.31 │   360750317729    360714345453   0.01 │  781016   781048  -0.00 │
│                      coq-verdi-raft │  540.60   542.24  -0.30 │  3762660087981   3762904937182  -0.01 │  828700   828924  -0.03 │
│                           coq-verdi │   45.40    45.53  -0.29 │   301849899803    301805026281   0.01 │  524216   523500   0.14 │
│                            coq-hott │  155.17   155.49  -0.21 │  1087365563153   1086981768206   0.04 │  463268   465568  -0.49 │
│                   coq-metacoq-utils │   23.02    23.06  -0.17 │   151456948683    151493051570  -0.02 │  602704   602708  -0.00 │
│               coq-mathcomp-analysis │  597.60   598.54  -0.16 │  4395948914394   4395881278577   0.00 │ 1790824  1793060  -0.12 │
│                 coq-category-theory │  603.71   604.62  -0.15 │  4495049203983   4494959674714   0.00 │  962148   961924   0.02 │
│                coq-mathcomp-algebra │  148.20   148.41  -0.14 │   999810478310    999889352851  -0.01 │ 1215804  1216244  -0.04 │
│                    coq-math-classes │   85.30    85.39  -0.11 │   521157252147    521153309842   0.00 │  505340   505588  -0.05 │
│               coq-mathcomp-fingroup │   26.39    26.41  -0.08 │   174320210827    174313985045   0.00 │  562032   561972   0.01 │
│        coq-fiat-crypto-with-bedrock │ 5304.03  5307.09  -0.06 │ 42971143506853  42977225767608  -0.01 │ 3245312  3245428  -0.00 │
│ coq-neural-net-interp-computed-lite │  233.47   233.56  -0.04 │  2259315759586   2259287452405   0.00 │  858564   860968  -0.28 │
│                         coq-bignums │   29.80    29.81  -0.03 │   189853052758    189883329521  -0.02 │  468576   469220  -0.14 │
│                       coq-equations │    8.20     8.20   0.00 │    53857489483     53867707019  -0.02 │  387612   387256   0.09 │
│                        coq-compcert │  301.72   301.45   0.09 │  1996825616244   1996835270044  -0.00 │ 1094300  1096272  -0.18 │
│             coq-metacoq-safechecker │  349.68   349.06   0.18 │  2679738513267   2678956524105   0.03 │ 1696324  1691564   0.28 │
│                  coq-metacoq-common │   67.00    66.86   0.21 │   438757744575    438893577217  -0.03 │ 1038976  1034524   0.43 │
│                      coq-coquelicot │   37.84    37.75   0.24 │   230675126701    230653876393   0.01 │  802824   804972  -0.27 │
│                coq-metacoq-template │  147.06   146.71   0.24 │   991331009956    994855624763  -0.35 │ 1039292  1037608   0.16 │
│               coq-engine-bench-lite │  129.48   129.17   0.24 │   969194845136    970764131532  -0.16 │ 1058352  1058240   0.01 │
│                      coq-test-suite │  502.78   501.41   0.27 │  3448425338917   3452977433420  -0.13 │ 2908556  2728076   6.62 │
│                          coq-stdlib │  346.52   345.52   0.29 │  1214569104924   1214585318946  -0.00 │  624816   623780   0.17 │
│                        coq-rewriter │  338.68   337.49   0.35 │  2538045635112   2537631567961   0.02 │ 1271036  1268896   0.17 │
│                            coq-corn │  677.62   675.14   0.37 │  4673665259344   4673209751020   0.01 │  705968   706028  -0.01 │
│                       coq-fiat-core │   57.75    57.51   0.42 │   347528530667    347554146024  -0.01 │  468304   470148  -0.39 │
│            coq-metacoq-translations │   16.63    16.56   0.42 │   121108656876    121390638597  -0.23 │  792528   789472   0.39 │
│         coq-rewriter-perf-SuperFast │  778.23   772.74   0.71 │  6035664923047   6035854605375  -0.00 │ 1364592  1364060   0.04 │
│                            coq-core │  131.29   130.30   0.76 │   539501520486    539162934652   0.06 │  479936   479176   0.16 │
│                    coq-fiat-parsers │  278.45   275.86   0.94 │  2147216592386   2146971340705   0.01 │ 2274380  2272848   0.07 │
│                       coq-fourcolor │ 1359.59  1345.67   1.03 │ 12564210465949  12564115377632   0.00 │  844864   841612   0.39 │
│              coq-mathcomp-ssreflect │  106.87   105.70   1.11 │   679602070735    679601033132   0.00 │ 1624520  1626608  -0.13 │
│                           coq-color │  245.94   243.23   1.11 │  1550860684426   1551086142878  -0.01 │ 1136448  1136720  -0.02 │
└─────────────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴─────────────────────────┘
~~~

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**